### ISSUE TITLE
BUG: datatetime not serializable in `to_json`

### ISSUE THREAD

**Issue by martinfleis** [geopandas/geopandas#1906]
Moved from https://github.com/martinfleis/geopandas-view/issues/49.

`datetime` columns raise an error in `to_json()` (see below). I guess we should follow pandas handling here https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_json.html.

```py
In [29]: nybb = gpd.read_file(gpd.datasets.get_path('nybb'))
    ...: nybb["time"] = [datetime.today() for ii in range(len(nybb.index))]

In [30]: nybb.to_json()
...
TypeError: Object of type Timestamp is not JSON serializable
```

**Comment by martinfleis** [comment]
The same error is shown for additional geometry columns. Those may be more tricky than the `datetime` case above.

**Comment by iboates** [comment]
What if additional geometry columns were converted to WKB or WKT strings? I would think that WKT is more appropriate, since JSON is a human readable format. But then again, a UNIX timestamp clearly prioritizes compactness over readability.

**Comment by iboates** [comment]
Also, slightly related (but probably warrants its own discussion), but I've been annoyed before when I wanted to export a GeoDataFrame to a file for very quick visual inspection in QGIS. Often, this is for colleagues who (despite my evangelical protestations) prefer shapefile. And so if there is a complex data type in the columns, like an additional geometry column, or a tuple column, or something. What if there was a `strict` parameter, which defaults to `True`, but when set to `False`, just outright ignores all columns which are considered unserializable in the output data format?

**Comment by martinfleis** [comment]
@iboates can you move the last bit to its own issue, please?

**Comment by iboates** [comment]
> 
> 
> @iboates can you move the last bit to its own issue, please?

Yeah, sorry I went off on a tangent there, I made it its own issue.

**Comment by jorisvandenbossche** [comment]
BTW, pandas actually serializes Timestamps as unix epoch numbers by default, but has a `date_format` keyword to opt for the ISO representation instead (only for the "table schema" format, ISO is used as default).

**Comment by iboates** [comment]
I took a very preliminary stab at this: https://github.com/geopandas/geopandas/pull/1920

I wanted to do a global map thing but I got tripped up with base types and implications about other complex types which may already be serializiable, and thought it might have unpredictable side-effects.

By default I assume we would cast timestamps to UNIX timestamps as per @jorisvandenbossche and WKB because it prioritizes brevity over readability which seems in line with how time is already handled in pandas `to_json()`

**Comment by oefe** [comment]
Well, as `to_json` has a `kwargs` that's passed into `json.dumps`, one can already serialize pandas columns containing timestamps (or other types that aren't JSON-serializable by default, such as geometries).

For example, passing a `default` arg for [`json.dumps`](https://docs.python.org/3/library/json.html#json.dump) to serialize the timestamp as epoch:

```python
def timestamp_encoder(obj):
    try:
        return obj.timestamp()
    except AttributeError:
        raise TypeError()

df.to_json(default=timestamp_encoder)
```

Or, using a [`JSONDecoder`](https://docs.python.org/3/library/json.html#json.JSONDecoder) to convert it to ISO format:
```python
import json
class TimestampEncoder(json.JSONEncoder):
    def default(self, obj):
        if isinstance(obj, pandas.Timestamp):
            return obj.isoformat()
        else:
            return json.JSONEncoder.default(self, obj)

df.to_json(cls=TimestampEncoder)
```

So instead of requiring a new conversion framework, one can simply use the existing conversion support in the standard library.

For geopandas-view (now `GeoDataFrame.explore`), this could simply enabled by adding a `json_kwds` argument that is passed to `json.dumps` in a similar way as the existing `*_kwds` arguments.

**Comment by asaph-marvin** [comment]
@martinfleis do you know if there is plan to have a fix for this bug? I currently have to manually change the datetime columns to strings, but I feel this should be handled without issues by `to_json` similarly to how pandas `to_json` works without added code.

**Comment by lexhuismans** [comment]
I am seeing a similar issue for ndarrays. 
```
TypeError: Object of type ndarray is not JSON serializable
```
I think arrays are converted to `ndarrays` by pandas but cannot be serialized after
